### PR TITLE
Drastically simplify distribution building for 2.11.

### PR DESCRIPTION
**Pull request by adriaanm**
See README and individual commits for info.

The main idea is that these scripts should do nothing more than repackaging org.scala-lang % scala-dist and its transitive dependencies. No duplication allowed. The process is reduced to pushing a tag to scala/scala-dist by the same name as the one from scala/scala (whose artifacts have been published to maven), and running `sbt s3-upload` on windows (for the msi) and unix (everything else).

We no longer package source code ourselves. GitHub provides source archives for every tag.

Review by @jsuereth, /cc @retronym, @gkossakowski.

This has all been tested locally. I'll do a dry run on jenkins tomorrow and link to the results.

### CODE REVIEW

**Comment by adriaanm** [comment]
This depends on scala/scala#3277 -- UPDATE: this has been merged, and jobs are successful, so full steam ahead!

**Comment by adriaanm** [comment]
Works on unix: https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x-unix/10/console
Works on windows: https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x-windows/6/console (the failure is an S3 failure also sometimes seen in the old jobs because of VirtualBox issues -- retrying as job 7)

**Comment by adriaanm** [comment]
The last thing between us and one-click releases for both maven and native packages is integrating sonatype staging repo closing (https://github.com/adriaanm/binfu/blob/master/sonafu.sh) into https://scala-webapps.epfl.ch/jenkins/view/scala-release-2.11.x/job/scala-release-2.11.x/.

For tying the binary compatibility knot (https://github.com/scala/scala-modules-build/blob/master/build.sh) I'll need to implement something very similar, so I'm hopeful we'll actually get there!

Of course, there's always more to be done. Then, we'll need automation for generating scala/scala-lang PRs (release notes, download pages,...) and a big red RELEASE button (which would release the closed sonatype repo).

File: README.md [hunk 1]
```diff
@@ -1,44 +1,36 @@
...
+as earlier ones did not publish the scala-dist artifact to maven.
 
-  * scala/scala must be in the user home directory
-  * Please run `ant dist-opt` in in the `scala/scala` project.
-  * zip the `<scala/scala>/dist/latest` directory into a file called `scala-dist.zip`
-  * Copy `scala-dist.zip` into `<scala/scala-dist>/target/tmp/scala-dist.zip`
-  * Start [sbt](https://github.com/harrah/xsbt) in the `<scala/scala-dist>` directory.
-  * Run one of the build commands.
+Due to limited resources, the native packages are quite rudimental,
+and tool-support isn't packaged at all.
```
Comment (jsuereth) [inline README.md:25]: That's probably ok.  maybe we can get some community contributions after this whole process is dramatically simplified.

File: build.sbt [hunk 2]
```diff
@@ -0,0 +1,28 @@
+import com.typesafe.sbt.SbtGit._
+import S3._
+
+// so we don't require a native git install
+useJGit
+
+// The version of this build determines the Scala version to package.
+// We look at the closest git tag that matches v[0-9].* to derive it.
+// For testing, the version may be overridden with -Dproject.version=...
+versionWithGit
```
Comment (jsuereth) [inline build.sbt:10]: I always smile when I see people using this.

File: project/Docs.scala [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ */
+object Docs {
+  import ScalaDist._
+
+  def settings: Seq[Setting[_]] = Seq(
+    name in UniversalDocs := s"scala-docs-${version.value}",
+    // libraryDependencies += scalaDistDep(version.value, "javadoc"), // seems not to be necessary
+    // need updateClassifiers to get javadoc jars
+    mappings in UniversalDocs ++= createMappingsWith(updateClassifiers.value.toSeq, universalDocsMappings)
```
Comment (jsuereth) [inline project/Docs.scala:22]: nice...

File: project/ScalaDist.scala [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                         distMappingGen: (ModuleID, Artifact, File) => Seq[(File, String)]): Seq[(File, String)] =
+    deps flatMap {
+      case d@(ScalaDistConfig, id, artifact, file) => distMappingGen(id, artifact, file)
+      case _ => Seq()
+    }
+
+  // used to make s3-upload upload the file produced by fileTask to the path scala/$version/${file.name}
+  private def uploadMapping(fileTask: TaskKey[File]) = Def.task {
+    val file = fileTask.value
+    file -> s"scala/${version.value}/${file.getName}"
```
Comment (jsuereth) [inline project/ScalaDist.scala:26]: nice simplification!

File: project/Unix.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ */
+object Unix {
+  val installTargetUnix     = SettingKey[File]("install-target-unix", "The location where we will install Scala.")
+  val installTargetUnixDocs = SettingKey[File]("install-target-docs-unix", "The location where we will install the Scala docs.")
+
+  def settings = Seq (
+    installTargetUnix     := file("/usr/share/scala"),
+    installTargetUnixDocs := file("/usr/share/doc/scala"),
+
+    (packageBin in Rpm) <<= ((packageBin in Rpm)
```
Comment (jsuereth) [inline project/Unix.scala]: Do you need it it to be staged?  I thought if you just et up the mappings it should work.

File: project/Unix.scala [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+    (packageBin in Debian) <<= ((packageBin in Debian)
+      dependsOn (stage in Universal)
+      dependsOn (stage in UniversalDocs)),
+
+    // symlinks for s"/usr/bin/$script" --> s"${installTargetUnix.value}/bin/$script"
+    linuxPackageSymlinks ++= (
+      (mappings in Universal).value collect {
+        case (file, name) if (name startsWith "bin/") && !(name endsWith ".bat") =>
+          LinuxSymlink("/usr/" + name, (installTargetUnix.value / name).getAbsolutePath)
```
Comment (jsuereth) [inline project/Unix.scala:31]: nice.  Is this borrowed from native packager?  Did the "GenericMappingsToLinux" settings not work?

File: project/Unix.scala [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+    rpmLicense     := Some("BSD"),
+    rpmGroup       := Some("Development/Languages"),
+
+    // This hack lets us ignore the RPM specific versioning junks.
+    packageBin in Rpm := {
+      val simplified = target.value / s"${(name in Rpm).value}-${version.value}.rpm"
+
+      val rpm = (packageBin in Rpm).value match {
+        case reported if reported.exists => reported
+        case _ => // hack on top of hack because RpmHelper.buildRpm is broken -- `spec.meta.arch` doesn't necessarily match the arch `rpmbuild` decided on
```
Comment (jsuereth) [inline project/Unix.scala]: This seems to be a MacOSX issue.  IF you think you know the issue, let me know. I'd love to fix it, but wasn't able to isolate the problem.  It isn't borked on RedHat/Ubuntu.

File: project/Unix.scala [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+      }
+
+      IO.copyFile(rpm, simplified)
+      simplified
+    },
+
+    // Debian Specific
+    name in Debian    := "scala",
+    debianPackageDependencies += "openjdk-6-jre | java6-runtime",
+    debianPackageDependencies += "libjansi-java",
```
Comment (jsuereth) [inline project/Unix.scala]: not really needed anymore.

File: project/Versioning.scala [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+    val Milestones = new scala.util.matching.Regex("(\\-\\d+)?\\-M(\\d+)")
+    val BuildNum = new scala.util.matching.Regex("\\-(\\d+)")
+
+    def calculateNumberFour(buildNum: Int = 0, rc: Int = 0, milestone: Int = 0) =
+      if(rc > 0 || milestone > 0) (buildNum)*400 + rc*20  + milestone
+      else (buildNum+1)*400 + rc*20  + milestone
+
+    version match {
+      case Majors(major, minor, bugfix, rest) => Option(rest) getOrElse "" match {
+        case Milestones(null, num)            => major + "." + minor + "." + bugfix + "." + calculateNumberFour(0,0,num.toInt)
```
Comment (jsuereth) [inline project/Versioning.scala]: Yay for ensuring windows upgrades work correctly.

File: project/Wix.scala [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+// can't call it Windows, that's a config name
+object Wix {
+  // Windows installer configuration
+  def settings: Seq[Setting[_]] = Seq(
+    mappings in Windows := (mappings in Universal).value,
+    // distributionFiles in Windows += (packageMsi in Windows).value,
+
+    wixProductId        := "7606e6da-e168-42b5-8345-b08bf774cb30",
+    wixProductUpgradeId := "6061c134-67c7-4fb2-aff5-32b01a186968",
```
Comment (jsuereth) [inline project/Wix.scala:18]: These are the same as before, right?

**Comment by jsuereth** [comment]
This looks very very good man!   I have one comment, one question, and otherwise looks great from me.

File: project/Wix.scala (hunk 10 above)
Comment (adriaanm) [inline project/Wix.scala:18]: yep

File: project/Unix.scala (hunk 6 above)
Comment (adriaanm) [inline project/Unix.scala:31]: It is borrowed -- sorry, hacked the last bits in anger, so didn't reuse all the bits I could've reused. Let's gradually refine from here.

File: project/Unix.scala (hunk 5 above)
Comment (adriaanm) [inline project/Unix.scala]: good point -- remnant from when we were still patching things; trying without and seems to going through
I'll add a commit

File: build.sbt (hunk 2 above)
Comment (adriaanm) [inline build.sbt:10]: :-) git ftw! I think we'll see tag-driven scala releases for 2.11.x!

File: project/Unix.scala (hunk 7 above)
Comment (adriaanm) [inline project/Unix.scala]: clarified mac issue in comment

**Comment by adriaanm** [comment]
Thanks!

**Comment by adriaanm** [comment]
@jsuereth, I believe I've addressed your feedback!

**Pull request merged**